interactive_templates/templates/v2/analysis/report_utils.py

```diff
@@ -1,10 +1,61 @@
 import json
 import re
+from pathlib import Path
 
 import matplotlib.dates as mdates
 import matplotlib.pyplot as plt
 
 
+BASE_DIR = Path(__file__).parents[1]
+OUTPUT_DIR = BASE_DIR / "output"
+ANALYSIS_DIR = BASE_DIR / "analysis"
+CODELIST_DIR = BASE_DIR / "codelists"
+
+
+BASE_DIR = Path(__file__).parents[1]
+OUTPUT_DIR = BASE_DIR / "output"
+ANALYSIS_DIR = BASE_DIR / "analysis"
+CODELIST_DIR = BASE_DIR / "codelists"
+
+
+def calculate_rate(df, value_col, population_col, rate_per=1000, round_rate=False):
+    """Calculates the number of events per 1,000 or passed rate_per variable of the population.
+    This function operates on the given measure table in-place, adding
+    a `rate` column.
+    Args:
+        df: A measure table.
+        value_col: The name of the numerator column in the measure table.
+        population_col: The name of the denominator column in the measure table.
+        rate_per: Value to calculate rate per
+        round: Bool indicating whether to round rate to 2dp.
+    """
+    if round_rate:
+        rate = round(df[value_col] / (df[population_col] / rate_per), 2)
+
+    else:
+        rate = df[value_col] / (df[population_col] / rate_per)
+
+    rate = rate.apply(round, ndigits=4)
+    return rate
+
+
+def relabel_sex(df):
+    sex_codes = {
+        "F": "Female",
+        "M": "Male",
+    }
+
+    df = df.replace({"sex": sex_codes})
+    return df
+
+
+def generate_expectations_codes(codelist, incidence=0.5):
+    expectations = {str(x): (1 - incidence) / 10 for x in codelist[0:10]}
+    # expectations = {str(x): (1-incidence) / len(codelist) for x in codelist}
```

Comment:
Delete commented line

Reply:
This doesn't actually do what it should for codelists with length <10. I'll update it